File: artifacts/ios-app/scripts/sync-app-icons.py
#!/usr/bin/env python3
"""Regenerate icon.png (native) and logo.png (in-app) from icon-source.png.

iOS push notifications use the native app icon — after updating assets, rebuild
with `pnpm mac:sim` (see AGENTS.md).
"""
from __future__ import annotations

from collections import deque
from pathlib import Path

from PIL import Image
import numpy as np

IMAGES = Path(__file__).resolve().parent.parent / "assets" / "images"
BRAND = (160, 194, 229)  # #A0C2E5


def load_source() -> Image.Image:
    for name in ("icon-source.png", "icon.png"):
        path = IMAGES / name
        if path.exists():
            return Image.open(path).convert("RGB")
    raise SystemExit(f"No source icon in {IMAGES}")


def main() -> None:
    src = load_source()
    arr = np.array(src, dtype=np.int16)
    h, w = arr.shape[:2]
    max_c = arr.max(axis=2)
    min_c = arr.min(axis=2)
    neutral_bright = ((max_c - min_c) < 12) & (max_c > 190)

    def flood(mask: np.ndarray) -> np.ndarray:
        visited = np.zeros_like(mask, dtype=bool)
        q: deque[tuple[int, int]] = deque()
        for x in range(w):
            for y in (0, h - 1):
                if mask[y, x]:
                    visited[y, x] = True
                    q.append((y, x))
        for y in range(h):
            for x in (0, w - 1):
                if mask[y, x] and not visited[y, x]:
                    visited[y, x] = True
                    q.append((y, x))
        while q:
            y, x = q.popleft()
            for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                ny, nx = y + dy, x + dx
                if 0 <= ny < h and 0 <= nx < w and mask[ny, nx] and not visited[ny, nx]:
                    visited[ny, nx] = True
                    q.append((ny, nx))
        return visited

    edge_bg = flood(neutral_bright)
    opaque = ~edge_bg

    def component_from(sy: int, sx: int, mask: np.ndarray) -> np.ndarray:
        comp = np.zeros_like(mask, dtype=bool)
        if not mask[sy, sx]:
            return comp
        q: deque[tuple[int, int]] = deque([(sy, sx)])
        comp[sy, sx] = True
        while q:
            y, x = q.popleft()
            for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                ny, nx = y + dy, x + dx
                if 0 <= ny < h and 0 <= nx < w and mask[ny, nx] and not comp[ny, nx]:
                    comp[ny, nx] = True
                    q.append((ny, nx))
        return comp

    logo_mask = component_from(h // 2, w // 2, opaque)
    rgba = np.zeros((h, w, 4), dtype=np.uint8)
    rgba[:, :, :3] = arr.astype(np.uint8)
    rgba[:, :, 3] = np.where(logo_mask, 255, 0).astype(np.uint8)
    logo_img = Image.fromarray(rgba, "RGBA")

    native = Image.new("RGB", (w, h), BRAND)
    native.paste(logo_img, (0, 0), logo_img)
    native.save(IMAGES / "icon.png", optimize=True)

    alpha = rgba[:, :, 3]
    ys, xs = np.where(alpha > 0)
    pad = 8
    x0, x1 = max(0, xs.min() - pad), min(w - 1, xs.max() + pad)
    y0, y1 = max(0, ys.min() - pad), min(h - 1, ys.max() + pad)
    logo_img.crop((x0, y0, x1 + 1, y1 + 1)).save(IMAGES / "logo.png", optimize=True)
    print(f"Wrote {IMAGES / 'icon.png'} and {IMAGES / 'logo.png'}")


if __name__ == "__main__":
    main()

File: artifacts/ios-app/scripts/sync-app-icons.mjs
#!/usr/bin/env node
/**
 * Regenerates icon.png (native) and logo.png (in-app) from icon-source.png.
 * Requires pngjs (installed with pnpm install).
 */
import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { PNG } from 'pngjs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const IMAGES = join(__dirname, '../assets/images');
const BRAND = { r: 160, g: 194, b: 229 };

function loadSource() {
  for (const name of ['icon-source.png', 'icon.png']) {
    const path = join(IMAGES, name);
    if (existsSync(path)) {
      return { path, png: PNG.sync.read(readFileSync(path)) };
    }
  }
  throw new Error(`No source icon in ${IMAGES}`);
}

function isNeutralBright(r, g, b) {
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  return max - min < 12 && max > 190;
}

function floodEdgeBackground(width, height, isBg) {
  const visited = new Uint8Array(width * height);
  const queue = [];

  const push = (x, y) => {
    const i = y * width + x;
    if (visited[i] || !isBg(i)) return;
    visited[i] = 1;
    queue.push(i);
  };

  for (let x = 0; x < width; x += 1) {
    push(x, 0);
    push(x, height - 1);
  }
  for (let y = 0; y < height; y += 1) {
    push(0, y);
    push(width - 1, y);
  }

  while (queue.length) {
    const i = queue.pop();
    const x = i % width;
    const y = (i - x) / width;
    if (x > 0) push(x - 1, y);
    if (x < width - 1) push(x + 1, y);
    if (y > 0) push(x, y - 1);
    if (y < height - 1) push(x, y + 1);
  }

  return visited;
}

function componentFromCenter(width, height, opaque) {
  const comp = new Uint8Array(width * height);
  const start = ((height / 2) | 0) * width + ((width / 2) | 0);
  if (!opaque[start]) return comp;

  const queue = [start];
  comp[start] = 1;

  while (queue.length) {
    const i = queue.pop();
    const x = i % width;
    const y = (i - x) / width;
    const neighbors = [];
    if (x > 0) neighbors.push(i - 1);
    if (x < width - 1) neighbors.push(i + 1);
    if (y > 0) neighbors.push(i - width);
    if (y < height - 1) neighbors.push(i + width);

    for (const n of neighbors) {
      if (!comp[n] && opaque[n]) {
        comp[n] = 1;
        queue.push(n);
      }
    }
  }

  return comp;
}

function bounds(comp, width, height) {
  let x0 = width;
  let y0 = height;
  let x1 = 0;
  let y1 = 0;

  for (let y = 0; y < height; y += 1) {
    for (let x = 0; x < width; x += 1) {
      if (!comp[y * width + x]) continue;
      x0 = Math.min(x0, x);
      y0 = Math.min(y0, y);
      x1 = Math.max(x1, x);
      y1 = Math.max(y1, y);
    }
  }

  return { x0, y0, x1, y1 };
}

function main() {
  const { path: sourcePath, png } = loadSource();
  const { width, height, data } = png;

  const isBg = (i) => {
    const r = data[i * 4];
    const g = data[i * 4 + 1];
    const b = data[i * 4 + 2];
    return isNeutralBright(r, g, b);
  };

  const edgeBg = floodEdgeBackground(width, height, isBg);
  const opaque = new Uint8Array(width * height);

  for (let i = 0; i < width * height; i += 1) {
    if (!edgeBg[i]) opaque[i] = 1;
  }

  const logo = componentFromCenter(width, height, opaque);
  const { x0, y0, x1, y1 } = bounds(logo, width, height);
  const pad = 8;
  const cx0 = Math.max(0, x0 - pad);
  const cy0 = Math.max(0, y0 - pad);
  const cx1 = Math.min(width - 1, x1 + pad);
  const cy1 = Math.min(height - 1, y1 + pad);
  const cw = cx1 - cx0 + 1;
  const ch = cy1 - cy0 + 1;

  const native = new PNG({ width, height });
  const cropped = new PNG({ width: cw, height: ch });

  for (let y = 0; y < height; y += 1) {
    for (let x = 0; x < width; x += 1) {
      const i = y * width + x;
      const di = i * 4;
      if (logo[i]) {
        native.data[di] = data[di];
        native.data[di + 1] = data[di + 1];
        native.data[di + 2] = data[di + 2];
        native.data[di + 3] = 255;
      } else {
        native.data[di] = BRAND.r;
        native.data[di + 1] = BRAND.g;
        native.data[di + 2] = BRAND.b;
        native.data[di + 3] = 255;
      }
    }
  }

  for (let y = 0; y < ch; y += 1) {
    for (let x = 0; x < cw; x += 1) {
      const sx = cx0 + x;
      const sy = cy0 + y;
      const si = (sy * width + sx) * 4;
      const di = (y * cw + x) * 4;
      if (logo[sy * width + sx]) {
        cropped.data[di] = data[si];
        cropped.data[di + 1] = data[si + 1];
        cropped.data[di + 2] = data[si + 2];
        cropped.data[di + 3] = 255;
      } else {
        cropped.data[di] = 0;
        cropped.data[di + 1] = 0;
        cropped.data[di + 2] = 0;
        cropped.data[di + 3] = 0;
      }
    }
  }

  writeFileSync(join(IMAGES, 'icon.png'), PNG.sync.write(native));
  writeFileSync(join(IMAGES, 'logo.png'), PNG.sync.write(cropped));
  console.log(`Synced icons from ${sourcePath}`);
}

main();
